### PR TITLE
iam: Phase I: API Key

### LINKED ISSUE #147: js: IAM support phase I (apikey)
Tracking: https://github.com/IBM-Cloud/gp-common/issues/24

This version will only use the gp specific `API-KEY` auth header, and not do any token management.
A future issue will handle token management and disable `API-KEY`.

### PR DESCRIPTION
**Pull request by srl295**
Fixes #147

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -54,11 +55,22 @@ class Client {
...
     this._options = options;
     if (!this._options.credentials) {
-      throw new Error("g11n-pipeline: missing 'credentials' " + Object.keys(Consts.exampleCredentials));
+      throw new Error("g11n-pipeline: missing 'credentials'"); // don't give specific contents
+      // could be CF or IAM.
     }
-    var missingField = utils.isMissingField(this._options.credentials, Object.keys(Consts.exampleCredentials));
-    if (missingField.length !== 0) {
-      throw new Error("g11n-pipeline: missing credentials fields: \"" + missingField.join(' ') + "\" - expected: " + Consts.exampleCredentialsString);
+    if(this._options.credentials.apikey) {
```
Comment (srl295) [inline lib/client.js]: if `apikey` is present, assume it is IAM.

File: lib/client.js [hunk 2]
```diff
@@ -106,16 +118,24 @@ class Client {
     const schemaUrl = this._schemaUrl = this._options.credentials.url + '/swagger.json';
     // if (debugREST) /*istanbul ignore next*/ console.log('.. fetching ' + schemaUrl);
 
-    const gphmac = new GpHmac("gp-hmac", this._options.credentials.userId, this._options.credentials.password);
-    if(this._options.basicAuth) {
-      // ignore basicAuth.
-      // throw Error('basicAuth is not supported'); // TODO: support this- maybe?
+    if (this._options.credentials.apikey) {
```
Comment (srl295) [inline lib/client.js]: if `apikey` is present, use the IAM interceptor

File: lib/gp-iam.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Generate HTTP Authorization header.
+ */
+GpIAM.prototype.apply = function(obj) {
+  if(this.VERBOSE) console.dir(obj, {color: true, depth: null});
+  if(obj.url.indexOf("/swagger.json") !== -1) return obj; // skip for swagger.json
+
+  const authHeader = this.API_KEY + ' ' + this.credentials.apikey;
+  if(this.VERBOSE)  console.log('hmacHeader = ' + authHeader);
+  obj.headers.Authorization = authHeader;
```
Comment (srl295) [inline lib/gp-iam.js]: this is the actual change.

**Comment by srl295** [comment]
note that actually `iam_endpoint` is not actually used, but it will be in the future ( #150 ), so I want to require/document it now.

File: lib/client.js (hunk 1 above)
Comment (sid4) [reply lib/client.js]: the assumption is correct

File: lib/gp-iam.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Generate HTTP Authorization header.
+ */
+GpIAM.prototype.apply = function(obj) {
+  if(this.VERBOSE) console.dir(obj, {color: true, depth: null});
+  if(obj.url.indexOf("/swagger.json") !== -1) return obj; // skip for swagger.json
+
+  const authHeader = this.API_KEY + ' ' + this.credentials.apikey;
+  if(this.VERBOSE)  console.log('hmacHeader = ' + authHeader);
```
Comment (yumaoka) [inline lib/gp-iam.js]: Do you really want `hmacHeader=` here?

Comment (sid4) [inline lib/gp-iam.js]: do we want to log actual API_KEY in console?

Comment (srl295) [reply lib/gp-iam.js]: only if VERBOSE is on.

Comment (srl295) [reply lib/gp-iam.js]: it's a debug comment, i can change it to `authHeader`

Comment (sid4) [reply lib/gp-iam.js]: I understand that, but I wasn't sure if it should ever be logged, like we never log it in our server side code. So just wanted to make sure it is intentional. I don't know the best practice for JS.

Comment (srl295) [reply lib/gp-iam.js]: I will change it to `Authorization: API-KEY ****` (using replace to obfuscate the actual apikey)

**Pull request merged**